File: lib/java/opentoken/src/main/java/com/truveta/opentoken/io/csv/PersonAttributesCSVWriter.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Map;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.io.PersonAttributesWriter;

/**
 * The PersonAttributeCSVWriter class is responsible for writing person
 * attributes to a CSV file.
 * It implements the {@link PersonAttributeWriter} interface.
 */
public class PersonAttributesCSVWriter implements PersonAttributesWriter {
    private static final Logger logger = LoggerFactory.getLogger(PersonAttributesCSVWriter.class);

    private final BufferedWriter fileWriter;
    private final CSVPrinter csvPrinter;
    private boolean headerWritten = false;

    /**
     * Initialize the class with the output file in CSV format.
     * 
     * @param filePath the output file path
     * @throws IOException if an I/O error occurs
     */
    public PersonAttributesCSVWriter(String filePath) throws IOException {
        fileWriter = new BufferedWriter(new FileWriter(filePath));
        // Use LF line endings to match Python output
        CSVFormat format = CSVFormat.Builder.create(CSVFormat.DEFAULT)
                .setRecordSeparator('\n')
                .get();
        csvPrinter = new CSVPrinter(fileWriter, format);
    }

    @Override
    public void writeAttributes(Map<String, String> data) {

        try {
            if (!headerWritten) {
                // Write the header
                csvPrinter.printRecord(data.keySet());
                headerWritten = true;
            }

            csvPrinter.printRecord(data.values());

        } catch (IOException e) {
            logger.error("Error in writing CSV file: {}", e.getMessage());
        }
    }

    @Override
    public void close() throws Exception {
        this.csvPrinter.close();
        this.fileWriter.close();
    }
}

File: lib/python/opentoken/src/main/opentoken/io/csv/person_attributes_csv_writer.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import csv
import logging
import os
from typing import Dict
from opentoken.io.person_attributes_writer import PersonAttributesWriter


logger = logging.getLogger(__name__)


class PersonAttributesCSVWriter(PersonAttributesWriter):
    """
    The PersonAttributeCSVWriter class is responsible for writing person
    attributes to a CSV file.
    It implements the PersonAttributesWriter interface.
    """

    def __init__(self, file_path: str):
        """
        Initialize the class with the output file in CSV format.

        Args:
            file_path: The output file path.

        Raises:
            IOError: If an I/O error occurs.
        """
        self.file_path = file_path

        # Create directory if it doesn't exist
        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        self.file_handle = open(file_path, 'w', newline='', encoding='utf-8')
        self.csv_writer = csv.writer(self.file_handle, lineterminator='\n')
        self.header_written = False

    def write_attributes(self, data: Dict[str, str]) -> None:
        """
        Write attributes to the CSV file.

        Args:
            data: A map of person attributes.
        """
        try:
            if not self.header_written:
                # Write the header
                self.csv_writer.writerow(data.keys())
                self.header_written = True

            self.csv_writer.writerow(data.values())

        except IOError as e:
            logger.error(f"Error in writing CSV file: {e}")
            raise

    def close(self) -> None:
        """Close the CSV writer and file handle."""
        if self.file_handle:
            self.file_handle.close()
